File: 2022/day04/python/part2.py
from pathlib import Path
file_path = Path(__file__, '..', '..', 'input.txt').resolve()

with open(file_path) as f:
    data = f.read().splitlines()

    overlaps = 0

    for line in data:
        elf1 = list(map(int, line.split(',')[0].split('-')))
        elf2 = list(map(int, line.split(',')[1].split('-')))

        if elf2[1] >= elf1[0] >= elf2[0] or elf2[0] <= elf1[1] <= elf2[1]:
            overlaps += 1
            print(line)

        elif elf1[0] <= elf2[0] <= elf2[1] and elf1[1] >= elf2[1] >= elf1[0]:
            overlaps += 1
            print(line)

    print(overlaps)

File: 2022/day05/javascript/part2.mjs
import { join } from "node:path";
import { readFile } from "node:fs/promises";

const filePath = new URL(join(import.meta.url, "..", "..", "input.txt"));

const file = await readFile(filePath, { encoding: "utf-8" });
const data = file.split("\n");
const stack = {};

let j = 0;

for (let [i, line] of data.entries()) {
  if (!isNaN(Number(String(line).trim()[0]))) {
    j = i + 2;
    break;
  }

  for (let j = 1; j < line.length; j += 4) {
    let k = (j % 4) + Math.floor(j / 4);
    if (String(line[j]).trim().length !== 0) {
      stack[k] = stack[k] ? stack[k] : [];
      stack[k].unshift(line[j]);
    }
  }
}

for (; j < data.length; j++) {
  const line = data[j];
  const count = isNaN(Number(line[5]))
    ? Number(line[5])
    : Number(line[5] + line[6]);

  const giver = Number(line[count > 9 ? 13 : 12]);
  const receiver = Number(line[line.length - 1]);

  const move = stack[giver].splice(stack[giver].length - count);
  stack[receiver].push(...move);
}

console.log(
  Object.values(stack)
    .map((item) => item[item.length - 1])
    .join("")
);
